range_queries/persistent_segment_tree_with_lazy_propagation.cpp

```diff
@@ -0,0 +1,323 @@
+/**
+ * @file
+ * @brief [Persistent segment tree with range updates (lazy
+ * propagation)](https://en.wikipedia.org/wiki/Persistent_data_structure)
+ * @details
+ * A normal segment tree facilitates making point updates and range queries in
+ * logarithmic time. Lazy propagation preserves the logarithmic time with range
+ * updates. So, a segment tree with lazy propagation enables doing range updates
+ * and range queries in logarithmic time, but it doesn't save any information
+ * about itself before the last update. A persistent data structure always
+ * preserves the previous version of itself when it is modified. That is, a new
+ * version of the segment tree is generated after every update. It saves all
+ * previous versions of itself (before every update) to facilitate doing range
+ * queries in any version. More memory is used ,but the logarithmic time is
+ * preserved because the new version points to the same nodes, that the previous
+ * version points to, that are not affected by the update. That is, only the
+ * nodes that are affected by the update and their ancestors are copied. The
+ * rest is copied using lazy propagation in the next queries. Thus preserving
+ * the logarithmic time because the number of nodes copied after any update is
+ * logarithmic.
+ *
+ * @author [Magdy Sedra](https://github.com/MSedra)
+ */
+#include <iostream>  /// for IO operations
+#include <memory>    /// to manage dynamic memory
+#include <vector>    /// for std::vector
+
+/**
+ * @namespace range_queries
+ * @brief Range queries algorithms
+ */
+namespace range_queries {
+
+/**
+ * @brief Range query here is range sum, but the code can be modified to make
+ * different queries like range max or min.
+ */
+class perSegTree {
+ private:
+    class Node {
+     public:
+        std::shared_ptr<Node> left;   /// pointer to the left node
+        std::shared_ptr<Node> right;  /// pointer to the right node
+        int64_t val, prop;  /// val is the value of the node (here equals to the
+                            /// sum of the leaf nodes children of that node),
+                            /// prop is the value to be propagated/added to all
+                            /// the leaf nodes children of that node
+
+        Node() {
+            val = prop = 0;
+            left = right = nullptr;
+        }
+    };
+
+    int n;  /// number of elements/leaf nodes in the segment tree
```

Comment:
We **always** got to initialize all variables, at least with `0`.
```suggestion
    int n = 0;  /// number of elements/leaf nodes in the segment tree
```